sematic/db/migrations/20230706150252_pre_alembic_parity.py

```diff
@@ -0,0 +1,339 @@
+                    original_run_id character(32),
+                    cache_key TEXT,
+                    user_id character(32) REFERENCES users(id),
+
+                    PRIMARY KEY (id),
+
+                    FOREIGN KEY(root_id) REFERENCES runs (id)
+                );
+                """
+            )
+
+            conn.execute("CREATE INDEX ix_runs_cache_key ON runs_new (cache_key);")
+            conn.execute(
+                "CREATE INDEX ix_runs_function_path ON runs_new (function_path);"
+            )
+
+            conn.execute(
+                """
+                INSERT INTO runs_new
+                    SELECT
+                        id,
+                        future_state,
+                        name,
+                        function_path,
+                        created_at,
+                        updated_at,
+                        started_at,
+                        ended_at,
+                        resolved_at,
+                        failed_at,
+                        parent_id,
+                        description,
+                        tags,
+                        source_code,
+                        root_id,
+                        nested_future_id,
+                        resource_requirements_json,
+                        exception_metadata_json,
+                        container_image_uri,
+                        external_exception_metadata_json,
+                        original_run_id,
+                        cache_key,
+                        user_id
+                    FROM runs;
+                """
+            )
+
+            conn.execute("DROP TABLE runs;")
+            conn.execute("ALTER TABLE runs_new RENAME TO runs;")
+
+            conn.execute(
+                """
+                CREATE TABLE artifacts_new (
+                    id character(40) NOT NULL,
+                    json_summary JSONB NOT NULL,
+                    created_at timestamp NOT NULL,
+                    updated_at timestamp NOT NULL,
+                    type_serialization JSONB NOT NULL,
+
+                    PRIMARY KEY (id)
+                );
+                """
+            )
+            conn.execute("INSERT INTO artifacts_new SELECT * FROM artifacts;")
+            conn.execute("DROP TABLE artifacts;")
+            conn.execute("ALTER TABLE artifacts_new RENAME TO artifacts;")
+
+            conn.execute(
+                """
+                CREATE TABLE resolutions_new (
+                    root_id TEXT NOT NULL,
+                    status TEXT NOT NULL,
+                    kind TEXT NOT NULL,
+                    container_image_uri TEXT,
+                    settings_env_vars JSONB NOT NULL,
+                    git_info_json JSONB,
+                    container_image_uris JSONB,
+                    client_version TEXT,
+                    cache_namespace TEXT,
+                    user_id character(32)
+                    REFERENCES users(id),
```

Comment:
Doesn't this need to be declared in the section below?

Reply:
> Doesn't this need to be declared in the section below?

fixed.

Reply:
> On the same subject, does metric_labels need to be redefined and copied as well, given that its primary key is defined inline, and not as a separate constraint?

i _think_ you're referring to the fact that it's currently written like so:

```
    metric_id TEXT NOT NULL PRIMARY KEY,
```

whereas there's the other syntax we use here:

```
    metric_id TEXT NOT NULL,
    PRIMARY KEY (metric_id)
```

both are valid syntax, and accomplish the same thing.  the latter one is the only valid syntax when your primary key spans multiple columns, but we don't have that here.  sqlite accepts both, so there's no need for additional work here.